datafusion/core/src/datasource/listing/table.rs

```diff
@@ -855,14 +855,17 @@ impl TableProvider for ListingTable {
         let input_partitions = input.output_partitioning().partition_count();
         let writer_mode = match self.options.insert_mode {
             ListingTableInsertMode::AppendToFile => {
-                if input_partitions > file_groups.len() {
+                if file_groups.is_empty() && self.options.single_file {
+                    // This is a hack, longer term append should be split out (#7994)
```

Comment:
Without this we get an error if this is the first write as there are no file_groups. Longer-term we shouldn't be conflating ListingTable IO and that of streaming use-cases